### ISSUE TITLE
Exception when using outbound SOAP channel with WS-Security

### ISSUE THREAD

**Issue by candlerb** [zatosource/zato#417]
Using an outbound SOAP channel with WS-Security authentication fails. The following exception is logged:

~~~
"/opt/zato/2.0.2/zato-server/src/zato/server/connection/http_soap/outgoing.py", 
line 139, in set_auth
self.soap[self.config['soap_version']]['header_template'].format(
AttributeError: 'HTTPSOAPWrapper' object has no attribute 'soap'
~~~

The problem is that `BaseHTTPSOAPWrapper.__init__` calls `set_auth` which in turn refers to `self.soap` if WSS authentication is selected. `self.soap` is initialized in the subclass `HTTPSOAPWrapper` but this is only done *after* calling the superclass `__init__`.

[Further details on mailing list](https://mailman-mail5.webfaction.com/pipermail/zato-discuss/2015-April/001080.html). This includes a simple workaround of moving the call to `__init__`. However it might be cleaner to move `set_auth` into the subclasses, especially since the existing `set_auth` logic behaves differently for HTTP_SOAP_SERIALIZATION_TYPE of suds versus string.

**Comment by candlerb** [comment]
Note: this issue also causes zato to fail to start up.

After upgrading to zato 2.0.3 (which removed my workaround), the server *appears* to start OK:

~~~
zato@cfprov:~/env/qs-1$ ./zato-qs-start.sh
Starting the Zato cluster quickstart-816651
Running sanity checks
[1/6] Redis connection OK
[2/6] SQL ODB connection OK
[3/6] Load-balancer started
[4/6] server1 started
[5/6] server2 started
[6/6] Web admin started
Zato cluster quickstart-816651 started
Visit https://zato.io/support for more information and support options
~~~

However, when I go into the web UI and try to list services, I get the following error:

~~~
Traceback (most recent call last):
File "/opt/zato/2.0.3/zato-web-admin/src/zato/admin/web/views/__init__.py", line 299, in __call__ response = self.invoke_admin_service()
File "/opt/zato/2.0.3/zato-web-admin/src/zato/admin/web/views/__init__.py", line 262, in invoke_admin_service return func(self.service_name, self.input)
File "/opt/zato/2.0.3/zato-web-admin/src/zato/admin/middleware.py", line 70, in invoke raise Exception(response.inner.text)
Exception: {"zato_env": {"details": "No server is available to handle the request", "result": "ZATO_ERROR", "cid": "K012345678901234567890123456"} }
~~~

I see the following processes running:

~~~
zato@cfprov:~/env/qs-1$ ps auxwww | grep zato
root     23057  0.0  0.1  65192  1500 pts/3    S    15:36   0:00 sudo su - zato
root     23058  0.0  0.1  64528  1132 pts/3    S    15:36   0:00 su - zato
zato     23059  0.0  0.3  22912  3488 pts/3    S    15:36   0:00 -su
zato     23127  2.2  8.2 221800 84352 pts/3    S    15:36   0:04 /opt/zato/2.0.3/bin/python /opt/zato/2.0.3/bin/py -m zato.agent.load_balancer.main /opt/zato/env/qs-1/load-balancer/config/repo
zato     23138  0.0  0.0  11960   944 ?        Ss   15:36   0:00 haproxy -D -f /opt/zato/env/qs-1/load-balancer/config/repo/zato.config -p /opt/zato/env/qs-1/load-balancer/zato-lb-agent.pid
zato     23169  3.9  9.1 234652 92932 pts/3    S    15:37   0:06 /opt/zato/2.0.3/bin/python /opt/zato/2.0.3/bin/py -m zato.admin.main
zato     23180  0.0  0.1  18452  1328 pts/3    R+   15:39   0:00 ps auxwww
zato     23181  0.0  0.0  11744   932 pts/3    S+   15:39   0:00 grep --color=auto zato
~~~

So the server1/server2 processes are not running. Checking in `server1/logs/server.log`:

~~~
...
2015-04-09 15:37:09,140 - INFO - 23153:Dummy-1 - gunicorn.main:22 - Booting worker with pid: 23153
2015-04-09 15:37:09,240 - INFO - 23154:Dummy-1 - gunicorn.main:22 - Booting worker with pid: 23154
2015-04-09 15:38:44,239 - ERROR - 23153:Dummy-1 - gunicorn.main:22 - Exception in worker process:
Traceback (most recent call last):
  File "/opt/zato/2.0.3/eggs/gunicorn-18.0-py2.7.egg/gunicorn/arbiter.py", line 494, in spawn_worker
    self.cfg.post_fork(self, worker)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/parallel.py", line 834, in post_fork
    ParallelServer.start_server(worker.app.zato_wsgi_app, arbiter.zato_deployment_key)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/parallel.py", line 801, in start_server
    parallel_server._after_init_accepted(server, zato_deployment_key)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/parallel.py", line 613, in _after_init_accepted
    self.worker_store.init()
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/worker.py", line 183, in init
    self.init_http_soap()
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/worker.py", line 314, in init_http_soap
    wrapper = self._http_soap_wrapper_from_config(config_dict.config)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/worker.py", line 274, in _http_soap_wrapper_from_config
    return HTTPSOAPWrapper(wrapper_config)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 164, in __init__
    super(HTTPSOAPWrapper, self).__init__(config, requests_module)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 65, in __init__
    self.set_auth()
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 139, in set_auth
    self.soap[self.config['soap_version']]['header_template'].format(
AttributeError: 'HTTPSOAPWrapper' object has no attribute 'soap'
Traceback (most recent call last):
  File "/opt/zato/2.0.3/eggs/gunicorn-18.0-py2.7.egg/gunicorn/arbiter.py", line 494, in spawn_worker
    self.cfg.post_fork(self, worker)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/parallel.py", line 834, in post_fork
    ParallelServer.start_server(worker.app.zato_wsgi_app, arbiter.zato_deployment_key)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/parallel.py", line 801, in start_server
    parallel_server._after_init_accepted(server, zato_deployment_key)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/parallel.py", line 613, in _after_init_accepted
    self.worker_store.init()
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/worker.py", line 183, in init
    self.init_http_soap()
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/worker.py", line 314, in init_http_soap
    wrapper = self._http_soap_wrapper_from_config(config_dict.config)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/worker.py", line 274, in _http_soap_wrapper_from_config
    return HTTPSOAPWrapper(wrapper_config)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 164, in __init__
    super(HTTPSOAPWrapper, self).__init__(config, requests_module)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 65, in __init__
    self.set_auth()
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 139, in set_auth
    self.soap[self.config['soap_version']]['header_template'].format(
AttributeError: 'HTTPSOAPWrapper' object has no attribute 'soap'
2015-04-09 15:38:44,249 - INFO - 23153:Dummy-1 - gunicorn.main:22 - Worker exiting (pid: 23153)
2015-04-09 15:38:44,266 - ERROR - 23153:Dummy-2 - springpython.context.ApplicationContext:22 - Could not destroy object 'parallel_server', exception 'Traceback (most recent call last):
  File "/opt/zato/2.0.3/eggs/springpython-1.3.0rc1-py2.7.egg/springpython/context/__init__.py", line 106, in shutdown_hook
    destroy_method()
  File "/opt/zato/2.0.3/zato-server/src/zato/server/base/parallel.py", line 859, in destroy
    self.broker_client.publish(msg, msg_type=msg_type)
AttributeError: 'ParallelServer' object has no attribute 'broker_client'
'
2015-04-09 15:39:15,295 - INFO - 23139:Dummy-1 - gunicorn.main:22 - Shutting down: Master
2015-04-09 15:39:15,296 - INFO - 23139:Dummy-1 - gunicorn.main:22 - Reason: Worker failed to boot.
~~~

So: apart from fixing the underlying bug with outgoing soap WSS, it might also be good if the `zato-qs-start.sh` script confirmed that the server process does not fail during boot.

**Comment by dsuch** [comment]
Done in 542234dd18e9ea68e4f0eac05b4676bbcaed6bd9 - also, backported to support/2.0. Will be included in 2.0.4.
